vmm/src/memory_manager.rs

```diff
@@ -252,45 +263,149 @@ impl MemoryManager {
+            }
+
+            let mut total_ram_size: u64 = 0;
+            for custom_region in custom_regions.iter() {
+                total_ram_size += custom_region.size;
+            }
+
+            arch_mem_regions = arch::arch_memory_regions(total_ram_size as GuestUsize);
+
+            let ram_regions: Vec<(GuestAddress, usize)> = arch_mem_regions
+                .iter()
+                .filter(|r| r.2 == RegionType::Ram)
+                .map(|r| (r.0, r.1))
+                .collect();
+
+            let mut custom_region = custom_regions.remove(0);
+            let mut custom_region_offset = 0;
+
+            for ram_region in ram_regions.iter() {
+                let mut ram_region_offset = 0;
+                let mut exit = false;
+
+                loop {
+                    let mut ram_region_consumed = false;
+                    let mut pull_next_custom_region = false;
+
+                    let ram_region_sub_size = ram_region.1 - ram_region_offset;
+                    let custom_region_sub_size = custom_region.size as usize - custom_region_offset;
+
+                    let file_offset = custom_region_offset as u64;
+                    let region_start = ram_region.0.unchecked_add(ram_region_offset as u64);
+                    let region_size = if custom_region_sub_size <= ram_region_sub_size {
+                        if custom_region_sub_size == ram_region_sub_size {
+                            ram_region_consumed = true;
+                        }
+
+                        ram_region_offset += custom_region_sub_size;
+                        pull_next_custom_region = true;
+
+                        custom_region_sub_size
+                    } else {
+                        custom_region_offset += ram_region_sub_size;
+                        ram_region_consumed = true;
+
+                        ram_region_sub_size
+                    };
+
+                    mem_regions.push(MemoryManager::create_ram_region(
+                        &custom_region.file,
+                        file_offset,
+                        region_start,
+                        region_size,
+                        false,
+                        prefault,
+                        custom_region.shared,
+                        custom_region.hugepages,
+                    )?);
+
+                    if pull_next_custom_region {
+                        // Get the next custom region and reset the offset.
+                        custom_region_offset = 0;
+                        if custom_regions.is_empty() {
+                            exit = true;
+                            break;
+                        }
+                        custom_region = custom_regions.remove(0);
+                    }
+
+                    if ram_region_consumed {
+                        break;
+                    }
+                }
+
+                if exit {
+                    break;
+                }
```

Comment:
Fitting custom regions into ram regions of various sizes seems to be inevitably convoluted. May be it is worth splitting this piece of code into a separate method, something like `fn create_memory_regions_from_custom_regions(&custom_regions, &ram_regions) -> Vector<memory_regions>`, and adding some comments to explain what it does in general.

Reply:
Yes, I will create two functions to reduce the complexity of the `MemoryManager::new()`.